File: frontend/src/lib/simulator/simulator.ts
// src/lib/simulator/simulator.ts
import { Cache } from './cache';
import { getPolicy } from './policies';
import type { CacheConfig, SimulationResult, AccessResult, CacheLineSnapshot } from './types';

export class CacheSimulator {
  config: CacheConfig;
  cache: Cache;
  policy: ReturnType<typeof getPolicy>;
  
  hits = 0;
  misses = 0;
  tick = 0;
  events: AccessResult[] = [];

  constructor(config: CacheConfig) {
    this.config = config;
    this.cache = new Cache(config.cacheSizeBytes, config.blockSizeBytes, config.associativity);
    this.policy = getPolicy(config.policy);
  }

  access(address: number): AccessResult {
    const { tag, setIndex, blockOffset } = this.cache.decodeAddress(address);
    const cset = this.cache.sets[setIndex];
    this.tick++;

    const hitWay = cset.findLine(tag);
    let hit = false;
    let targetWay = 0;
    let evictedTag: number | null = null;

    if (hitWay !== null) {
      // HIT
      this.hits++;
      hit = true;
      targetWay = hitWay;
      this.policy.onAccess(cset, hitWay, this.tick);
    } else {
      // MISS
      this.misses++;
      const emptyWay = cset.hasEmptyLine();

      if (emptyWay !== null) {
        targetWay = emptyWay;
      } else {
        targetWay = this.policy.chooseVictim(cset);
        evictedTag = cset.lines[targetWay].tag;
      }

      // Load block
      const line = cset.lines[targetWay];
      line.valid = true;
      line.tag = tag;
      line.frequency = 0;
      line.lastUsed = 0;
      line.insertOrder = 0;
      
      this.policy.onFill(cset, targetWay, this.tick);
    }

    // Capture snapshot for visualization (deep copy needed for sets/lines we care about,
    // but to save memory, we only take a snapshot of valid lines)
    const snapshot: CacheLineSnapshot[] = [];
    for (let si = 0; si < this.cache.sets.length; si++) {
      const s = this.cache.sets[si];
      for (let wi = 0; wi < s.lines.length; wi++) {
        const l = s.lines[wi];
        if (l.valid) {
          snapshot.push({
            set: si,
            way: wi,
            valid: l.valid,
            tag: l.tag,
            frequency: l.frequency,
            lastUsed: l.lastUsed,
          });
        }
      }
    }

    const event: AccessResult = {
      tick: this.tick,
      address,
      tag,
      setIndex,
      blockOffset,
      hit,
      way: targetWay,
      evictedTag,
      cacheSnapshot: snapshot,
    };

    this.events.push(event);
    return event;
  }

  run(addresses: number[]): SimulationResult {
    for (let i = 0; i < addresses.length; i++) {
      this.access(addresses[i]);
    }

    const totalAccesses = this.hits + this.misses;
    const hitRate = totalAccesses > 0 ? this.hits / totalAccesses : 0;
    const missRate = totalAccesses > 0 ? this.misses / totalAccesses : 0;

    const amat = this.config.hitTime + missRate * this.config.missPenalty;
    const cpi = this.config.baseCpi + missRate * this.config.missPenalty;

    return {
      metrics: {
        totalAccesses,
        hits: this.hits,
        misses: this.misses,
        hitRate,
        missRate,
        amat,
        cpi,
      },
      trace: this.events,
      organization: this.cache.getOrganization(),
      numSets: this.cache.numSets,
      associativity: this.cache.associativity,
      blockOffsetBits: this.cache.blockOffsetBits,
      indexBits: this.cache.indexBits,
    };
  }
}

// Helper to run simulation completely asynchronously so it doesn't block UI
export function runSimulationAsync(addresses: number[], config: CacheConfig): Promise<SimulationResult> {
  return new Promise((resolve) => {
    // In a real huge-scale app, we'd use a Web Worker.
    // For now, a setTimeout pushes it to the macro-task queue, 
    // letting React render loading states first.
    setTimeout(() => {
      const sim = new CacheSimulator(config);
      const res = sim.run(addresses);
      resolve(res);
    }, 50);
  });
}

File: backend/simulator/simulator.py
"""
simulator.py - The main CacheSimulator engine.
Processes a list of memory addresses, records every access event,
and produces per-step trace data plus aggregate metrics.
"""
from __future__ import annotations
from dataclasses import dataclass, field
from typing import List, Optional
from .cache import Cache, CacheLine
from .policies import get_policy
from .metrics import SimulationMetrics


@dataclass
class AccessEvent:
    """Represents the outcome of a single memory access."""
    tick: int
    address: int
    tag: int
    set_index: int
    block_offset: int
    hit: bool
    evicted_tag: Optional[int]
    way: int
    cache_snapshot: list = field(default_factory=list)


class CacheSimulator:
    """
    Drives a Cache through a sequence of addresses, applying a replacement policy.

    Parameters
    ----------
    cache_size_bytes : int
    block_size_bytes : int
    associativity    : int
    policy           : str  ("LRU" | "FIFO" | "LFU")
    hit_time         : float  (cycles)
    miss_penalty     : float  (cycles)
    base_cpi         : float
    """

    def __init__(
        self,
        cache_size_bytes: int = 1024,
        block_size_bytes: int = 64,
        associativity: int = 1,
        policy: str = "LRU",
        hit_time: float = 1.0,
        miss_penalty: float = 100.0,
        base_cpi: float = 1.0,
    ) -> None:
        self.cache = Cache(cache_size_bytes, block_size_bytes, associativity)
        self.policy = get_policy(policy)
        self.hit_time = hit_time
        self.miss_penalty = miss_penalty
        self.base_cpi = base_cpi
        self.hits = 0
        self.misses = 0
        self.tick = 0
        self.events: List[AccessEvent] = []

    def access(self, address: int) -> AccessEvent:
        """Process one memory access and return the access event."""
        tag, set_index, block_offset = self.cache.decode_address(address)
        cset = self.cache.get_set(set_index)
        self.tick += 1

        hit_way = cset.find_line(tag)
        evicted_tag: Optional[int] = None

        if hit_way is not None:
            # ---- HIT ----
            self.hits += 1
            self.policy.on_access(cset, hit_way, self.tick)
            event = AccessEvent(
                tick=self.tick,
                address=address,
                tag=tag,
                set_index=set_index,
                block_offset=block_offset,
                hit=True,
                evicted_tag=None,
                way=hit_way,
                cache_snapshot=self.cache.state_snapshot(),
            )
        else:
            # ---- MISS ----
            self.misses += 1
            empty_way = cset.has_empty_line()

            if empty_way is not None:
                target_way = empty_way
            else:
                target_way = self.policy.choose_victim(cset)
                evicted_tag = cset.lines[target_way].tag

            # Load new block
            line = cset.lines[target_way]
            line.valid = True
            line.tag = tag
            line.data = address  # symbolic
            line.frequency = 0
            line.last_used = 0
            line.load_order = 0
            self.policy.on_fill(cset, target_way, self.tick)

            event = AccessEvent(
                tick=self.tick,
                address=address,
                tag=tag,
                set_index=set_index,
                block_offset=block_offset,
                hit=False,
                evicted_tag=evicted_tag,
                way=target_way,
                cache_snapshot=self.cache.state_snapshot(),
            )

        self.events.append(event)
        return event

    def run(self, addresses: List[int]) -> None:
        """Process a full list of addresses."""
        for addr in addresses:
            self.access(addr)

    def get_metrics(self) -> SimulationMetrics:
        return SimulationMetrics(
            total_accesses=self.hits + self.misses,
            hits=self.hits,
            misses=self.misses,
            hit_time=self.hit_time,
            miss_penalty=self.miss_penalty,
            base_cpi=self.base_cpi,
        )

    def get_trace(self) -> List[dict]:
        """Return the event log as a serializable list."""
        return [
            {
                "tick": e.tick,
                "address": f"0x{e.address:08X}",
                "address_dec": e.address,
                "tag": e.tag,
                "set_index": e.set_index,
                "block_offset": e.block_offset,
                "result": "HIT" if e.hit else "MISS",
                "evicted_tag": e.evicted_tag,
                "way": e.way,
            }
            for e in self.events
        ]

    def get_step_snapshots(self) -> List[dict]:
        """Return per-step cache snapshots for animated playback."""
        return [
            {
                "tick": e.tick,
                "result": "HIT" if e.hit else "MISS",
                "address": f"0x{e.address:08X}",
                "set_index": e.set_index,
                "way": e.way,
                "evicted_tag": e.evicted_tag,
                "snapshot": e.cache_snapshot,
            }
            for e in self.events
        ]
